localstack/services/stepfunctions/asl/component/state/state_execution/state_task/service/state_task_service_ecs.py

```diff
@@ -0,0 +1,127 @@
+            raw_parameters[self._STARTED_BY_PARAMETER_RAW_KEY] = self._STARTED_BY_PARAMETER_VALUE
+        super()._before_eval_execution(
+            env=env, resource_runtime_part=resource_runtime_part, raw_parameters=raw_parameters
+        )
+
+    def _eval_service_task(
+        self,
+        env: Environment,
+        resource_runtime_part: ResourceRuntimePart,
+        normalised_parameters: dict,
+    ):
+        service_name = self._get_boto_service_name()
+        api_action = self._get_boto_service_action()
+        ecs_client = boto_client_for(
+            region=resource_runtime_part.region,
+            account=resource_runtime_part.account,
+            service=service_name,
+        )
+        response = getattr(ecs_client, api_action)(**normalised_parameters)
+        response.pop("ResponseMetadata", None)
+
+        # AWS outputs the description of the task, not the output of run_task.
+        match self._get_boto_service_action():
+            case "run_task":
+                self._normalise_response(response=response, service_action_name="run_task")
+                cluster_arn: str = response["Tasks"][0]["ClusterArn"]
+                task_arn: str = response["Tasks"][0]["TaskArn"]
+                describe_tasks_output = ecs_client.describe_tasks(
+                    cluster=cluster_arn, tasks=[task_arn]
+                )
+                describe_tasks_output.pop("ResponseMetadata", None)
+                self._normalise_response(
+                    response=describe_tasks_output, service_action_name="describe_tasks"
+                )
+                env.stack.append(describe_tasks_output)
+                return
+
+        env.stack.append(response)
+
+    def _sync_to_run_task(
+        self,
+        env: Environment,
+        resource_runtime_part: ResourceRuntimePart,
+    ) -> None:
+        ...
+        ecs_client = boto_client_for(
+            region=resource_runtime_part.region,
+            account=resource_runtime_part.account,
+            service="ecs",
+        )
+        submission_output: dict = env.stack.pop()
+        task_arn: str = submission_output["Tasks"][0]["TaskArn"]
+        cluster_arn: str = submission_output["Tasks"][0]["ClusterArn"]
+
+        def _has_terminated() -> Optional[dict]:
+            describe_tasks_output = ecs_client.describe_tasks(cluster=cluster_arn, tasks=[task_arn])
+            last_status: str = describe_tasks_output["tasks"][0]["lastStatus"]
+
+            if last_status == "STOPPED":
+                self._normalise_response(
+                    response=describe_tasks_output, service_action_name="describe_tasks"
+                )
+                return describe_tasks_output["Tasks"][0]  # noqa
+
+            return None
+
+        termination_output: Optional[dict] = None
+        while env.is_running() and termination_output is None:
+            termination_output: Optional[dict] = _has_terminated()
```

Comment:
does the env.is_running cause a wait interval between the describe_tasks checks or is this a kind of busy waiting scheme?

Reply:
This logic simply terminates the evaluation if the program execution received a stop signal whilst synching. As sync may be executed until the state timeout, this just clears resources sooner by exiting the sync now rather than at timeout.

Reply:
I might not have expressed my question correctly. What is preventing this from doing basically unlimited describe_tasks calls in the while loop, is there any throttling/minimum interval involved?

Reply:
Unlimited requests are prevented by stop signals or the state's timeout, however yes absolutely, good point, we don't have throttling and we should add one.

Reply:
I have added a temporary solution to this by invoking a throttling function in each iteration. We should probably abstract this logic in the callback class to enforce all sync operations to implement an sync iteration pattern which is automatically throttled and interrupted.